[Site Isolation] Main frame navigation to a site loaded in a child frame should not reuse message receivers

charliewolfe opened the pull request:
#### cbe8887ec47918a004b8b38dbd9efb1977b0b3c7
<pre>
[Site Isolation] Main frame navigation to a site loaded in a child frame should not reuse message receivers
<a href="https://bugs.webkit.org/show_bug.cgi?id=287450">https://bugs.webkit.org/show_bug.cgi?id=287450</a>
<a href="https://rdar.apple.com/144577663">rdar://144577663</a>

Reviewed by Alex Christensen.

This is work towards enabling the back/forward cache.

Currently, if the main frame navigates to a site that is already loaded in a child frame,
ProvisionalPageProxy reuses the remote page’s message receiver and page identifier. This cannot work when
we need to cache the iframe’s page that is being navigated away from. Reusing the message receiver and
page’s main frame is only necessary when a new window is opened, so we should restrict this behavior to
that case.

* Source/WebKit/UIProcess/BrowsingContextGroup.cpp:
(WebKit::BrowsingContextGroup::removeFrameProcess):
(WebKit::BrowsingContextGroup::addRemotePage):
(WebKit::BrowsingContextGroup::removePage):
* Source/WebKit/UIProcess/BrowsingContextGroup.h:
* Source/WebKit/UIProcess/ProvisionalPageProxy.cpp:
(WebKit::ProvisionalPageProxy::ProvisionalPageProxy):
(WebKit::ProvisionalPageProxy::initializeWebPage):
(WebKit::ProvisionalPageProxy::didFailProvisionalLoadForFrame):
If a provisional navigation in the main frame fails, we need to re-add the RemotePageProxy that was
removed from the browsing context. We cannot use transitionProvisionalPageToRemotePage if the
ProvisionalPageProxy did not take the message receiver from the taken RemotePageProxy.

* Source/WebKit/UIProcess/ProvisionalPageProxy.h:
* Source/WebKit/UIProcess/RemotePageProxy.cpp:
(WebKit::RemotePageProxy::~RemotePageProxy):
Moving the Close IPC message to the RemotePageProxy destructor fixes an issue where the RemotePageProxy
taken by ProvisionalPageProxy would be destroyed without being evicted from the web process. With this
change, removePageFromProcess() is no longer needed.

(WebKit::RemotePageProxy::removePageFromProcess): Deleted.
* Source/WebKit/UIProcess/RemotePageProxy.h:

Canonical link: <a href="https://commits.webkit.org/290219@main">https://commits.webkit.org/290219@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/84293d4c2d2aa07db336eba7ce052124a5fd9fec

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/89258 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/8782 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/44090 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/94241 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/40019 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/9170 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/17074 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/68765 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/26437 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/92260 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/7031 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/81003 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/49126 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/6780 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/39126 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/77139 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/36371 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/96073 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/16438 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/12112 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/77644 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/16694 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/76790 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/76939 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/21361 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/19959 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/9558 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/14001 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/16452 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/21763 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/16193 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/19644 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/17974 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebKit/UIProcess/ProvisionalPageProxy.cpp:
```diff
@@ -439,7 +442,10 @@ void ProvisionalPageProxy::didFailProvisionalLoadForFrame(FrameInfoData&& frameI
+        if (m_isProcessSwappingForNewWindow)
+            protectedBrowsingContextGroup()->transitionProvisionalPageToRemotePage(*this, Site(request.url()));
+        else if (m_takenRemotePage)
+            protectedBrowsingContextGroup()->addRemotePage(*page, *WTFMove(m_takenRemotePage));
```
releaseNonNull()

achristensen07 commented on Source/WebKit/UIProcess/BrowsingContextGroup.cpp:
```diff
@@ -153,12 +152,18 @@ void BrowsingContextGroup::addPage(WebPageProxy& page)
+{
+    auto it = m_remotePages.find(page);
+    if (it == m_remotePages.end())
+        return;
```
If we don't have a HashSet here, wouldn't we want to add one?  I'd expect a call to WeakHashMap::ensure or an ASSERT_NOT_REACHED because there's something else ensuring this won't happen.

charliewolfe committed:
[Site Isolation] Main frame navigation to a site loaded in a child frame should not reuse message receivers
https://bugs.webkit.org/show_bug.cgi?id=287450
rdar://144577663

Reviewed by Alex Christensen.

This is work towards enabling the back/forward cache.

Currently, if the main frame navigates to a site that is already loaded in a child frame,
ProvisionalPageProxy reuses the remote page’s message receiver and page identifier. This cannot work when
we need to cache the iframe’s page that is being navigated away from. Reusing the message receiver and
page’s main frame is only necessary when a new window is opened, so we should restrict this behavior to
that case.

* Source/WebKit/UIProcess/BrowsingContextGroup.cpp:
(WebKit::BrowsingContextGroup::removeFrameProcess):
(WebKit::BrowsingContextGroup::addRemotePage):
(WebKit::BrowsingContextGroup::removePage):
* Source/WebKit/UIProcess/BrowsingContextGroup.h:
* Source/WebKit/UIProcess/ProvisionalPageProxy.cpp:
(WebKit::ProvisionalPageProxy::ProvisionalPageProxy):
(WebKit::ProvisionalPageProxy::initializeWebPage):
(WebKit::ProvisionalPageProxy::didFailProvisionalLoadForFrame):
If a provisional navigation in the main frame fails, we need to re-add the RemotePageProxy that was
removed from the browsing context. We cannot use transitionProvisionalPageToRemotePage if the
ProvisionalPageProxy did not take the message receiver from the taken RemotePageProxy.

* Source/WebKit/UIProcess/ProvisionalPageProxy.h:
* Source/WebKit/UIProcess/RemotePageProxy.cpp:
(WebKit::RemotePageProxy::~RemotePageProxy):
Moving the Close IPC message to the RemotePageProxy destructor fixes an issue where the RemotePageProxy
taken by ProvisionalPageProxy would be destroyed without being evicted from the web process. With this
change, removePageFromProcess() is no longer needed.

(WebKit::RemotePageProxy::removePageFromProcess): Deleted.
* Source/WebKit/UIProcess/RemotePageProxy.h:

Canonical link: https://commits.webkit.org/290219@main
